### ISSUE TITLE
add a simple test framework

### ISSUE THREAD

**Issue by Pomax** [mozilla/thimble.mozilla.org#1606]
thimble currently has no tests beyond "does npm install pass", so adding a simple test framework that is invoked with `npm test` is a great idea.

It doesn't even need to be a complex monster like mocha with phantom (even though it's less of a monster than it sounds like), it could even just be a simple tests.js file in vanilla nodejs JavaScript that starts the server, tries to access a few URLs, and then shuts the server off again, with a process.exit(1) when something happened that shouldn't have happened.

**Comment by rajpatel507** [comment]
can I take this one? 
I have integrated jasmine with phantomjs without any glue code(changed jasmine boot file) with one of my project. do you want something like that?

According to your idea I am confuse about how can test.js file will know about something went wrong or not?

**Comment by rajpatel507** [comment]
can I take this one? 
I have integrated jasmine with phantomjs without any glue code(changed jasmine boot file) with one of my project. do you want something like that?

According to your idea I am confuse about how can test.js file will know about something went wrong or not?

**Comment by humphd** [comment]
@Pomax do you want to clarify your intent here a bit for @rajpatel507's sake?

**Comment by Pomax** [comment]
Certainly, sorry for the delay - there are several types of testing that are needed:

- unit testing of code (e.g. verifying that functions and sets of functions do what they should do, both on expected and unexpected input). This would be a *lot* of work to set up and not something we should start with
- integration testing (e.g. making sure that code in thimble that relies on brackets, id, etc. communicates properly). This will *also* be a lot of work.
- automated user testing: this is the jasmine/karma/phantom/etc approach, and the one we want to focus on first.

The elaborate on the automated user testing: jasmine/karma/phantom is reasonably easy to set up, http://orizens.com/wp/topics/my-setup-for-testing-js-with-jasmine-karma-phantomjs-angularjs/ is a decent read (although of course we're not using angular) and explains how to set up the basics.

Before we start testing that "things go wrong", we want to set up code that lets us *do* automated user testing at all: rather than tests that help us flag problems, we can write a few tests of which we know what the result would be: clicking the "start from scratch" button on the homepage should start thimble. Inside thimble, clicking the project title should change the UI to one where focus is on an input field that lets us type a new title, doing so and hitting "enter" should have changed title, that sort of thing. 

The important part is that we first set up the framework for testing, before we start writing actually meaningful tests: it's always a good idea to split up work into smaller bits so that each patch you write only adds a small bit of new functionality, rather than doing everything in a single huge effort. Adding a test framework and writing all the tests for unit, integration, and automated user testing as a single PR would be impossible to review, but writing a small patch for "just adding the test framework dependencies with basic npm scripts for running off of the most minimal test.js" is easy to do, easy to review, and easy to land before we then move on to building out the battery of tests.

**Comment by rajpatel507** [comment]
@Pomax Before we go into integration part we need to choose automated test framework, Jasmine is default choice for most of JS developer for its easy to lean and use approach but since then lots of new JS framework have come so someone with good experience in automaton testing can suggest good framework and then we can start integrating it with project.

**Comment by Pomax** [comment]
Never make sweeping statements about "the JS community", there are too many things we do to make any claims about what "we" do as a collective.

Let's start with a basic karma/phantom setup: as first step, we'll need to add karma/chai/mocha/phantom to the dependencies list, and set up a basic karma conf and simple test.js file that doesn't do much more than just assert that it can run at all.
